axlearn/common/checkpointer_orbax_emergency_replicator.py

```diff
@@ -0,0 +1,548 @@
+
+    def wait_until_finished(self):
+        self._manager.wait_until_finished()
+
+    def restore_from_dir(
+        self,
+        step: int,
+        state: Union[Nested[Tensor], Nested[TensorSpec]],
+        *,
+        ckpt_dir: str,
+        validation: CheckpointValidationType = CheckpointValidationType.EXACT,
+    ) -> Nested[Tensor]:
+        spec = self._get_spec(step=step, state=state)
+        logging.info("Restoring TF savables from directory %s", ckpt_dir)
+        check_state_structure(
+            read_index_file(ckpt_dir), target_structure=spec["index"], validation=validation
+        )
+        restore_tf_savables(
+            spec["tf_ckpt_map"], dir=os.path.join(ckpt_dir, f"tf_{jax.process_index()}")
+        )
+        multihost_utils.sync_global_devices(ckpt_dir)
+        return state
+
+    def stop(self):
+        self._executor.shutdown(wait=True)
+
+
+def _wait_for_file_to_disappear(f, timeout=300):
+    for _ in range(timeout):
+        if not f.exists():
+            return True
+        time.sleep(1)
+    return False
+
+
+def _extract_step(f):
+    # The base file name is formatted as {job_name}-s{step}-n{node_rank}-w{worker_rank}
+    return f.rsplit("-", 3)[1][1:]
+
+
+def _block_and_process_restore_dir(directory, timeout=300):
+    """Block until a file ending with `.restore` appears, then extract the step number and rename
+    the directory using the step number.
+    """
+    suffix = ".restore"
+    for _ in range(timeout):
+        files = os.listdir(directory)
+        for f in files:
+            if f.endswith(suffix):
+                step = _extract_step(f)
+                if step != "0":
+                    os.rename(epath.Path(directory) / f, epath.Path(directory) / step)
+                    logging.info(
+                        "Renamed restore directory at step %s to %s.",
+                        step,
+                        epath.Path(directory) / step,
+                    )
+                else:
+                    logging.info("Found a restore directory at step 0, skipping renaming.")
+                return
+        time.sleep(1)
+    logging.info("%d seconds have passed but no .restore file was found.", timeout)
+
+
+def _retrieve_jax_init_info(local_ckpt_dir):
+    """Retrieve JAX init info from a local file."""
+    jax_init_info_file = "jax-init-info.txt"
+    local_jax_init_info_file = epath.Path(local_ckpt_dir) / jax_init_info_file
+    # Allow time for the JAX init info file to be populated by GKE.
+    # File only populated when the worker with process id of 0 is determined.
+    for i in range(900):
+        if local_jax_init_info_file.exists():
+            return local_jax_init_info_file.read_text().split("\n")[:2]
+        logging.info(
+            "Unable to locate %s after %d seconds, sleeping for 1 second before retrying...",
+            jax_init_info_file,
+            i,
+        )
+        time.sleep(1)
+    logging.info(
```

Comment:
I suggest throwing exception here instead, as nothing good will come out of this if neither processId nor coordinator's IP is known